### PR TITLE
feat: implemented openRouter-pkce

### PR DESCRIPTION
**Pull request by Chigala**
### Description

This pull request implements the integration of OpenRouter into the application. The main feature added is a new button that allows users to select the OpenRouter provider and get their API key from openRouter. 

### Changes Made

1. Added a new button to the user interface, allowing users to choose the OpenRouter provider.
2. Implemented the button click event to redirect users to OpenRouter's authorization page.
3. Added logic to retrieve the API key from OpenRouter after the user grants access.
4. Injected the obtained API key into an input field, enabling users to seamlessly use OpenRouter's services.

### How It Works

1. When a user selects the "OpenRouter" provider, the "OpenRouter" button becomes visible.
2. Upon clicking the button, users are redirected to OpenRouter's authorization page.
3. Once the user grants access, the application automatically retrieves the API key from OpenRouter.
4. The obtained API key is then injected into an input field on the user interface.
5. This implementation eliminates the need for users to manually copy and paste the API key.

### Additional Notes

- Error handling has been thoroughly addressed to ensure a smooth user experience.
- The implementation adheres to best practices and security measures while interacting with OpenRouter's service.

### Testing

Extensive testing has been performed to verify the functionality of the OpenRouter integration. Test cases include:
- Button visibility and functionality
- Successful redirection to OpenRouter's authorization page
- Retrieval and injection of the API key into the input field

### Screenshots

image

Please review the changes and provide any feedback or suggestions for improvements. Thank you!

Chigala

### CODE REVIEW

File: src/lib/ai.ts [hunk 1]
```diff
@@ -249,3 +249,12 @@ export const calculateTokenCost = (tokens: number, model: ChatCraftModel) => {
   console.warn(`Unknown pricing for model ${model.toString()}`);
   return 0;
 };
+
+export const openRouterPkceWindow = () => {
+  let callbackUrl = "https://chatcraft.org";
+  if (!process.env.NODE_ENV || process.env.NODE_ENV === "development") {
+    callbackUrl = "http://localhost:5173";
```
Comment (tarasglek) [inline src/lib/ai.ts:256]: this is wrong, use window.location instead of hardcoding callback url

File: src/router.tsx [hunk 2]
```diff
@@ -10,9 +10,14 @@ export default createBrowserRouter([
...
   {
     path: "/",
-    async loader() {
+    async loader({ request }) {
+      const openRouterCode = new URL(request.url).searchParams.get("code");
       try {
         const recentChat = await db.chats.orderBy("date").last();
+        //if there is a recent chat and a code, redirect to the chat with the code
+        if (recentChat && openRouterCode) {
+          return redirect(`/c/${recentChat.id}?code=${openRouterCode}`);
```
Comment (tarasglek) [inline src/router.tsx:19]: what's the purpose of this code?

**Comment by tarasglek** [comment]
Please add equivalent functionality to Settings dialog

File: src/lib/ai.ts (hunk 1 above)
Comment (Chigala) [reply src/lib/ai.ts:256]: Ohh,Thanks. I didn't think of that. I'll implement this

**Comment by Chigala** [comment]
Ohk, I'll check that out

**Comment by tarasglek** [comment]
@humphd can you comment on these changes? Is this ok to land

File: src/hooks/use-codeExchange.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { useState, useEffect } from "react";
```
[the remark this answers is not in the document]
Comment (Chigala) [reply src/hooks/use-codeExchange.ts:1]: I'll integrate the logic into the use-settings hooks instead. Then strip the url of the code. Thanks

File: src/router.tsx (hunk 2 above)
Comment (Chigala) [reply src/router.tsx:19]: Noted. Thanks

File: src/components/Message/AppMessage/Instructions.tsx [hunk 4]
```diff
@@ -101,6 +100,11 @@ function Instructions(props: MessageBaseProps) {
                 Save
               </Button>
             </Flex>
+            {provider === "OpenRouter.ai" && (
+              <Button mt="3" size="sm" onClick={openRouterPkceWindow}>
+                Get key from OpenRouter{" "}
```
Comment (humphd) [inline src/components/Message/AppMessage/Instructions.tsx:105]: s/key/API Key/

File: src/components/PreferencesModal.tsx [hunk 5]
```diff
@@ -204,6 +204,11 @@ function PreferencesModal({ isOpen, onClose, finalFocusRef }: PreferencesModalPr
                 value={settings.apiKey || ""}
                 onChange={(e) => setSettings({ ...settings, apiKey: e.target.value })}
               />
+              {provider === "OpenRouter.ai" && !settings.apiKey && (
+                <Button mt="3" size="sm" onClick={openRouterPkceWindow}>
+                  Get key from OpenRouter{" "}
```
Comment (humphd) [inline src/components/PreferencesModal.tsx:209]: Same there, s/key/API Key/

File: src/hooks/use-settings.tsx [hunk 6]
```diff
@@ -31,8 +31,48 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
     serializer,
     deserializer,
   });
+
   const [state, setState] = useState<Settings>(settings || defaults);
 
+  //get code from the url
```
Comment (humphd) [inline src/hooks/use-settings.tsx:37]: Let's be more clear.  This `code` is for OpenRouter.ai, so let's name it that, and indicate this in the comment.  Also, let's cache the URL params object, so we can work with it later. 

```js
// Get the OpenRouter.ai code, if present on the URL
const searchParams = new URLSearchParams(window.location.search);
const openRouterCode = searchParams.get("code");
```

File: src/hooks/use-settings.tsx [hunk 7]
```diff
@@ -31,8 +31,48 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
...
+  const code = new URLSearchParams(window.location.search).get("code");
+
+  // const handleSetApiKey = (apiKey: string) => {
+  //   setApiKeyValue(apiKey);
+  // };
+
+  const handleCodeExchange = (code: string) => {
```
Comment (humphd) [inline src/hooks/use-settings.tsx:44]: `handleOpenRouterCodeExchange`

File: src/hooks/use-settings.tsx [hunk 8]
```diff
@@ -31,8 +31,48 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
...
+      .then((response) => response.json())
+      .then((data) => {
+        const apiKey = data.key;
+        if (apiKey !== undefined) {
+          setSettings({ ...state, apiKey: apiKey });
+          // Strip out code from the URL
+          const urlWithoutCode = window.location.href.split("?")[0];
+          window.history.replaceState({}, document.title, urlWithoutCode);
+        } else {
+          console.error("Error: API key is undefined.");
```
Comment (humphd) [inline src/hooks/use-settings.tsx:62]: "OpenRouter OAuth PKCE response missing API key"

File: src/hooks/use-settings.tsx [hunk 9]
```diff
@@ -31,8 +31,48 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
     serializer,
     deserializer,
   });
+
   const [state, setState] = useState<Settings>(settings || defaults);
 
+  //get code from the url
+  const code = new URLSearchParams(window.location.search).get("code");
+
+  // const handleSetApiKey = (apiKey: string) => {
```
Comment (humphd) [inline src/hooks/use-settings.tsx:40]: Remove this

File: src/hooks/use-settings.tsx [hunk 10]
```diff
@@ -31,8 +31,48 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
...
+      method: "POST",
+      body: JSON.stringify(requestBody),
+    })
+      .then((response) => response.json())
+      .then((data) => {
+        const apiKey = data.key;
+        if (apiKey !== undefined) {
+          setSettings({ ...state, apiKey: apiKey });
+          // Strip out code from the URL
+          const urlWithoutCode = window.location.href.split("?")[0];
```
Comment (humphd) [inline src/hooks/use-settings.tsx:59]: This is going to lose data if we ever start relying on the query params later.  I would do something like this:

```js
searchParams.delete('code');
const { origin, pathname, hash } = window.location;
const urlWithoutCode = `${origin}${pathname}${hash}${searchParams.toString()}`
```

File: src/lib/ai.ts [hunk 11]
```diff
@@ -249,3 +249,10 @@ export const calculateTokenCost = (tokens: number, model: ChatCraftModel) => {
   console.warn(`Unknown pricing for model ${model.toString()}`);
   return 0;
 };
+
+export const openRouterPkceWindow = () => {
```
Comment (humphd) [inline src/lib/ai.ts:253]: `openRouterPkceRedirect` would make more sense

File: src/lib/ai.ts [hunk 12]
```diff
@@ -249,3 +249,10 @@ export const calculateTokenCost = (tokens: number, model: ChatCraftModel) => {
...
+  const callbackUrl = window.location.href;
+  const hostname = new URL(callbackUrl).origin;
```
Comment (humphd) [inline src/lib/ai.ts:255]: `window.location` is already a `URL`, why do this again?  Also, why call an `origin` a `hostname`? What I have you before was more clear.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/components/Message/AppMessage/Instructions.tsx (hunk 4 above)
Comment (Chigala) [reply src/components/Message/AppMessage/Instructions.tsx:105]: Sorry, I don't quite get this.

File: src/hooks/use-settings.tsx (hunk 6 above)
Comment (Chigala) [reply src/hooks/use-settings.tsx:37]: Noted

File: src/hooks/use-settings.tsx (hunk 7 above)
Comment (Chigala) [reply src/hooks/use-settings.tsx:44]: Noted

File: src/hooks/use-settings.tsx (hunk 8 above)
Comment (Chigala) [reply src/hooks/use-settings.tsx:62]: I'll add this

File: src/hooks/use-settings.tsx (hunk 9 above)
Comment (Chigala) [reply src/hooks/use-settings.tsx:40]: Noted

File: src/components/Message/AppMessage/Instructions.tsx (hunk 4 above)
Comment (Chigala) [reply src/components/Message/AppMessage/Instructions.tsx:105]: Ohh,ohk.

File: src/lib/ai.ts (hunk 12 above)
Comment (humphd) [reply src/lib/ai.ts:255]: No problem, thanks for working on this.  Ping me when it's ready to review again.

File: src/hooks/use-settings.tsx (hunk 6 above)
Comment (Chigala) [reply src/hooks/use-settings.tsx:37]: Also, where should I cache the URL params object to?

Comment (humphd) [reply src/hooks/use-settings.tsx:37]: Exactly what I did in my code (i.e., create a variable here, so we can use it below):

```ts
const searchParams = new URLSearchParams(window.location.search);
```

Comment (Chigala) [reply src/hooks/use-settings.tsx]: Ohk

**Review by humphd — APPROVED** [review]
Code looks good to me.  @tarasglek can you test with your actual OpenRouter account?  If this works OK for you, I'd say it's good to land.

Nice job @Chigala!

**Comment by tarasglek** [comment]
It's been working since first push. Only concern was re code quality.
Thanks for getting that sorted to both of you!

**Pull request merged**